### PR TITLE
Load functions without guice

### PR DESCRIPTION
**Pull request by mfussenegger**
This will make it possible to use Functions within guice modules like
the `MetadataModule`, to use them in `fromXContent` implementations like
`ForeignTable.fromXContent` - to parse expressions into symbols.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/plugins/CompositeClassLoader.java [hunk 1]
```diff
@@ -47,10 +47,14 @@ protected Class<?> findClass(String name) throws ClassNotFoundException {
...
-     */
-    public static ExtendedPluginsClassLoader create(ClassLoader parent, List<ClassLoader> extendedLoaders) {
-        return new ExtendedPluginsClassLoader(parent, extendedLoaders);
+    @Override
+    public Enumeration<URL> getResources(String name) throws IOException {
+        List<URL> urls = new ArrayList<>();
+        super.getResources(name).asIterator().forEachRemaining(urls::add);
+        for (var loader : loaders) {
+            Enumeration<URL> resources = loader.getResources(name);
+            resources.asIterator().forEachRemaining(urls::add);
```
Comment (matriv) [inline server/src/main/java/org/elasticsearch/plugins/CompositeClassLoader.java:56]: Shouldn't we break the `loaders` iteration if the resource is loaded?

**Review by matriv — APPROVED** [review]
Thank you!
Left a minor question, and I'd wait for @seut to also have a look.

File: server/src/main/java/org/elasticsearch/plugins/CompositeClassLoader.java (hunk 1 above)
Comment (mfussenegger) [reply server/src/main/java/org/elasticsearch/plugins/CompositeClassLoader.java:56]: No, this should return a enumeration of URLs over all loaders, not just the first.

What might be worth doing is using a set to de-duplicate entries.

File: server/src/main/java/io/crate/metadata/Functions.java [hunk 2]
```diff
@@ -69,9 +74,39 @@ public Functions copyOf() {
...
     }
 
-    @Inject
-    public Functions(Map<FunctionName, List<FunctionProvider>> functionImplementationsBySignature) {
-        this.functionImplementations = functionImplementationsBySignature;
+    public static class Builder {
+
+        private HashMap<FunctionName, List<FunctionProvider>> providersByName = new HashMap<>();
+
+        public void put(Signature signature, FunctionFactory factory) {
```
Comment (seut) [inline server/src/main/java/io/crate/metadata/Functions.java:81]: Afaik this won't replace a factory for the same signature? If so, wouldn't `put` be a bit misleading?

```suggestion
        public void add(Signature signature, FunctionFactory factory) {
```

**Review by seut — APPROVED** [review]
Nice! Just a more minor interface naming comment

File: server/src/main/java/io/crate/metadata/Functions.java (hunk 2 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/metadata/Functions.java]: Good point. Renamed it.

**Pull request merged**